node/actors/executor/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+use storage::{FallbackReplicaStateStore, ReplicaStateStore, WriteBlockStore};
+use sync_blocks::SyncBlocks;
+use utils::pipe;
+
+mod config;
+mod io;
+mod metrics;
+#[cfg(test)]
+mod tests;
+
+pub use self::config::{ConsensusConfig, ExecutorConfig, GossipConfig};
+
+/// Validator-related part of [`Executor`].
+#[derive(Debug)]
+struct ValidatorExecutor {
+    /// Consensus network configuration.
+    config: ConsensusConfig,
+    /// Validator key.
+    key: validator::SecretKey,
+    /// Store for replica state.
+    replica_state_store: Arc<dyn ReplicaStateStore>,
+    /// Sender of blocks finalized by the consensus algorithm.
+    blocks_sender: channel::UnboundedSender<FinalBlock>,
+}
+
+impl ValidatorExecutor {
+    /// Returns consensus network configuration.
+    fn consensus_config(&self) -> network::consensus::Config {
+        network::consensus::Config {
+            // Consistency of the validator key has been verified in constructor.
+            key: self.key.clone(),
+            public_addr: self.config.public_addr,
+        }
+    }
+}
+
+/// Executor allowing to spin up all actors necessary for a consensus node.
+#[derive(Debug)]
+pub struct Executor<S> {
+    /// General-purpose executor configuration.
+    executor_config: ExecutorConfig,
+    /// Secret key of the node.
+    node_key: node::SecretKey,
+    /// Block and replica state storage used by the node.
+    storage: Arc<S>,
+    /// Validator-specific node data.
+    validator: Option<ValidatorExecutor>,
+}
+
+impl<S: WriteBlockStore + 'static> Executor<S> {
+    /// Creates a new executor with the specified parameters.
+    pub fn new(
+        node_config: ExecutorConfig,
+        node_key: node::SecretKey,
+        storage: Arc<S>,
+    ) -> anyhow::Result<Self> {
+        anyhow::ensure!(
+            node_config.gossip.key == node_key.public(),
+            "config.gossip.key = {:?} doesn't match the secret key {:?}",
+            node_config.gossip.key,
+            node_key
+        );
+
+        Ok(Self {
+            executor_config: node_config,
+            node_key,
+            storage,
+            validator: None,
+        })
+    }
+
+    /// Sets validator-related data for the executor.
+    pub fn set_validator(
+        &mut self,
+        config: ConsensusConfig,
+        key: validator::SecretKey,
+        replica_state_store: Arc<dyn ReplicaStateStore>,
+        blocks_sender: channel::UnboundedSender<FinalBlock>,
+    ) -> anyhow::Result<()> {
+        let public = &config.key;
+        anyhow::ensure!(
+            *public == key.public(),
+            "config.consensus.key = {public:?} doesn't match the secret key {key:?}"
+        );
+        anyhow::ensure!(
+            self.executor_config
+                .validators
+                .iter()
+                .any(|validator_key| validator_key == public),
+            "Key {public:?} is not a validator per validator set {:?}",
+            self.executor_config.validators
```

Comment:
Having a validator config, even though you are not a validator currently should be a valid setup. Once the validator set is dynamic, a node should be able to join and leave the consensus network. For now it would be enough to just ignore the ConsensusNetwork config if node is not a validator.